core/lib/zksync_core/src/state_keeper/io/seal_logic/l2_block_seal_subtasks.rs

```diff
@@ -0,0 +1,584 @@
+use std::sync::Arc;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use futures::FutureExt;
+use tokio::task::JoinHandle;
+use zksync_dal::{Core, CoreDal};
+use zksync_db_connection::{connection::Connection, connection_pool::ConnectionPool};
+use zksync_types::{event::extract_added_tokens, L2BlockNumber};
+
+use crate::state_keeper::{
+    metrics::{L2BlockSealStage, L2_BLOCK_METRICS},
+    updates::L2BlockSealCommand,
+};
+
+/// Helper struct that encapsulates parallel l2 block sealing logic.
+#[derive(Debug)]
+pub(crate) struct L2BlockSealProcess;
+
+impl L2BlockSealProcess {
+    pub fn all_subtasks() -> Vec<Box<dyn L2BlockSealSubtask>> {
+        vec![
+            Box::new(MarkTransactionsInL2BlockSubtask),
+            Box::new(InsertStorageLogsSubtask),
+            Box::new(InsertFactoryDepsSubtask),
+            Box::new(InsertTokensSubtask),
+            Box::new(InsertEventsSubtask),
+            Box::new(InsertL2ToL1LogsSubtask),
+        ]
+    }
+
+    pub fn subtasks_len() -> u32 {
+        Self::all_subtasks().len() as u32
+    }
+
+    pub async fn run_subtasks(
+        command: Arc<L2BlockSealCommand>,
+        pool: ConnectionPool<Core>,
+    ) -> anyhow::Result<()> {
+        let subtasks = Self::all_subtasks();
+        let handles: Vec<_> = subtasks
+            .into_iter()
+            .map(|subtask| {
+                subtask
+                    .spawn(command.clone(), pool.clone())
+                    .map(|res| res.unwrap_or_else(|err| Err(err.into())))
+            })
+            .collect();
```

Comment:
Nit: Since `try_join_all()` accepts iterator inputs, you don't need to `collect` handles.